{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyevolve import G2DList, G1DList, Mutators\n",
    "from pyevolve import GAllele, Initializators, GSimpleGA\n",
    "# from utils import load_data\n",
    "from expert_system import ExpertSystem\n",
    "from utils import load_data, expected_profit_campaign_predicted, expected_profit_campaign_actual\n",
    "from utils import expected_profit_campaign_predicted_actual, matches_count\n",
    "from prediction_system import PredictionModel\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "init_cutpoints = {\n",
    "        \"age\": [30, 40, 50],\n",
    "        \"income\": [2500, 5000, 7500],\n",
    "        \"avbal\": [14000, 20000, 27000],\n",
    "        \"avtrans\": [1000, 1500, 2400],\n",
    "        \"cip\": [3, 5, 7]\n",
    "    }\n",
    "df = load_data(\"data/custdatabase.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_list(cutpoints):\n",
    "    list_ = []\n",
    "    for key, value in cutpoints.items():\n",
    "        list_ += value\n",
    "    return list_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_dict(chromosome):\n",
    "    dict_ = {}\n",
    "    keys = init_cutpoints.keys()\n",
    "    i = 0\n",
    "    for key in keys:\n",
    "        dict_[key] = chromosome[i:i+3]\n",
    "        i += 3\n",
    "    return dict_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training classifier.\n",
      "Training complete.\n",
      "Gen. 0 (0.00%): Max/Min/Avg Fitness(Raw)             [99.52(101.00)/40.84(37.08)/82.93(82.93)]\n",
      "Gen. 1 (0.20%): Max/Min/Avg Fitness(Raw)             [81.92(100.00)/38.89(0.00)/68.27(68.27)]\n",
      "\n",
      "\tA break was detected, you have interrupted the evolution !\n",
      "\n",
      "Gen. 1 (0.20%): Max/Min/Avg Fitness(Raw)             [81.92(100.00)/38.89(0.00)/68.27(68.27)]\n",
      "Total time elapsed: 195.844 seconds.\n",
      "- GenomeBase\n",
      "\tScore:\t\t\t 100.000000\n",
      "\tFitness:\t\t 81.921380\n",
      "\n",
      "\tParams:\t\t {'allele': - GAlleles\n",
      "\tHomogeneous:\t False\n",
      "\tList size:\t 15\n",
      "\tAlleles:\n",
      "\n",
      "Allele for position 0:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [18] to [89]\n",
      "\n",
      "Allele for position 1:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [18] to [89]\n",
      "\n",
      "Allele for position 2:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [18] to [89]\n",
      "\n",
      "Allele for position 3:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [356] to [19955]\n",
      "\n",
      "Allele for position 4:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [356] to [19955]\n",
      "\n",
      "Allele for position 5:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [356] to [19955]\n",
      "\n",
      "Allele for position 6:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [88] to [88682]\n",
      "\n",
      "Allele for position 7:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [88] to [88682]\n",
      "\n",
      "Allele for position 8:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [88] to [88682]\n",
      "\n",
      "Allele for position 9:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [8] to [8789]\n",
      "\n",
      "Allele for position 10:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [8] to [8789]\n",
      "\n",
      "Allele for position 11:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [8] to [8789]\n",
      "\n",
      "Allele for position 12:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [0] to [10]\n",
      "\n",
      "Allele for position 13:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [0] to [10]\n",
      "\n",
      "Allele for position 14:\n",
      "- GAlleleRange\n",
      "\tReal:\t\t False\n",
      "\tRanges Count:\t 1\n",
      "\tRange List:\n",
      "\t\t\t Range from [0] to [10]\n",
      "\n",
      "}\n",
      "\n",
      "\tSlot [Evaluator] (Count: 1)\n",
      "\t\tName: eval_func - Weight: 0.50\n",
      "\tSlot [Initializator] (Count: 1)\n",
      "\t\tName: t_init - Weight: 0.50\n",
      "\tSlot [Mutator] (Count: 1)\n",
      "\t\tName: G1DListMutatorAllele - Weight: 0.50\n",
      "\t\tDoc:  The mutator of G1DList, Allele Mutator\n",
      "\n",
      "    To use this mutator, you must specify the *allele* genome parameter with the\n",
      "    :class:`GAllele.GAlleles` instance.\n",
      "\n",
      "    \n",
      "\tSlot [Crossover] (Count: 1)\n",
      "\t\tName: G1DListCrossoverSinglePoint - Weight: 0.50\n",
      "\t\tDoc:  The crossover of G1DList, Single Point\n",
      "\n",
      "    .. warning:: You can't use this crossover method for lists with just one element.\n",
      "\n",
      "    \n",
      "\n",
      "- G1DList\n",
      "\tList size:\t 15\n",
      "\tList:\t\t [30, 40, 50, 2500, 5000, 7500, 14000, 20000, 27000, 1000, 1500, 2400, 3, 5, 7]\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "chromosome_ = convert_to_list(init_cutpoints)\n",
    "\n",
    "setOfAlleles = GAllele.GAlleles()\n",
    "keys = init_cutpoints.keys()\n",
    "for key in keys:\n",
    "    for _ in range(0, 3):\n",
    "        if key != 'cip':\n",
    "            a = GAllele.GAlleleRange(int(df[key].min()), int(df[key].max()))\n",
    "            setOfAlleles.add(a)\n",
    "        else:\n",
    "            a = GAllele.GAlleleRange(0, 10)\n",
    "            setOfAlleles.add(a)\n",
    "\n",
    "df_actual = pd.read_csv(\"data/Cust_Actual.csv\")\n",
    "df_actual_ = pd.read_csv(\"data/Cust_Actual.csv\", index_col=['index'])\n",
    "\n",
    "genome = G1DList.G1DList(len(chromosome_))\n",
    "genome.setParams(allele=setOfAlleles)\n",
    "\n",
    "def check_sorted(chromosome):\n",
    "    i = 0\n",
    "    for _ in range(5):\n",
    "        original = chromosome[i:i+3]\n",
    "        sorted_ = sorted(original)\n",
    "        if (original != sorted_):\n",
    "            return False\n",
    "        i += 3\n",
    "    return True\n",
    "\n",
    "model = PredictionModel()\n",
    "\n",
    "def eval_func(chromosome):\n",
    "    cutpoints = convert_to_dict(chromosome)\n",
    "    if check_sorted(chromosome):\n",
    "        expert = ExpertSystem(df, cutpoints)\n",
    "        expected_profit_campaign_pred, predicted_index = expected_profit_campaign_predicted(model, expert, df)\n",
    "        expected_profit_campaign_act, actual_index = expected_profit_campaign_actual(df_actual)\n",
    "        expected_profit_campaign_pred_act = expected_profit_campaign_predicted_actual(df_actual_, predicted_index)\n",
    "        matches = matches_count(actual_index, predicted_index)\n",
    "        return matches + 100*(abs(expected_profit_campaign_pred-expected_profit_campaign_act)) / expected_profit_campaign_act\n",
    "#         print(matches)\n",
    "#         return matches\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def t_init(genome, **args):\n",
    "    genome.genomeList = chromosome_\n",
    "    \n",
    "#t_init(genome)\n",
    "\n",
    "genome.evaluator.set(eval_func)\n",
    "genome.mutator.set(Mutators.G1DListMutatorAllele)\n",
    "genome.initializator.set(t_init)\n",
    "\n",
    "# Genetic Algorithm Instance\n",
    "ga = GSimpleGA.GSimpleGA(genome)\n",
    "ga.setPopulationSize(100)\n",
    "ga.setGenerations(500)\n",
    "\n",
    "# Do the evolution, with stats dump\n",
    "# frequency of 10 generations\n",
    "ga.evolve(freq_stats=1)\n",
    "\n",
    "# Best individual\n",
    "print(ga.bestIndividual())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
